chain/beacon/sync_manager.go

```diff
@@ -0,0 +1,294 @@
+				s.log.Debug("sync_manager", "request already filled", "last", last.Round, "request", request.upTo)
+				continue
+			}
+			// check if it's been a while we haven't received a new round from
+			// sync. Either there is a sync in progress but it's stuck, so we
+			// quit it and start a new one, or there isn't and we start one.
+			// We always give a delay of a few periods since the one next to "now"
+			// might not be exactly ready yet so only after a few periods we know we
+			// must have gotten some data.
+			upperBound := lastRoundTime + int(s.period.Seconds())*s.factor
+			if upperBound < int(s.clock.Now().Unix()) {
+				// we haven't received a new block in a while
+				// -> time to start a new sync
+				cancel()
+				lastCtx, cancel = context.WithCancel(context.Background())
+				go s.sync(lastCtx, request)
+			}
+		case <-s.newSync:
+			// just received a new beacon from sync, we keep track of this time
+			lastRoundTime = int(s.clock.Now().Unix())
+		case <-s.done:
+			s.log.Info("sync_manager", "exits")
+			cancel()
+			return
+		}
+	}
+}
+
+func (s *SyncManager) sync(ctx context.Context, request requestInfo) {
+	s.log.Debug("sync_manager", "starting new sync", "up_to", request.upTo, "nodes", peersToString(request.nodes))
+	// shuffle through the nodes
+	for _, n := range rand.Perm(len(request.nodes)) {
+		node := request.nodes[n]
+		if s.tryNode(ctx, request.upTo, node) {
+			return
+		}
+	}
+	s.log.Debugw("Tried all nodes without success", "sync_manager", "failed sync")
+}
+
+// tryNode tries to sync up with the given peer up to the given round, starting
+// from the last beacon in the store. It returns true if the objective was
+// reached (store.Last() returns upTo) and false otherwise.
+func (s *SyncManager) tryNode(global context.Context, upTo uint64, n net.Peer) bool {
+	// we put a cancel to still keep the global context open but stop with this
+	// peer if things go sideway
+	cnode, cancel := context.WithCancel(global)
+	defer cancel()
+	last, err := s.store.Last()
+	if err != nil {
+		return false
+	}
+	beaconCh, err := s.client.SyncChain(cnode, n, &proto.SyncRequest{
+		FromRound: last.Round + 1,
+	})
+	if err != nil {
+		s.log.Debug("sync_manager", "unable_to_sync", "with_peer", n.Address(), "err", err)
+		return false
+	}
+
+	s.log.Debug("sync_manager", "start_follow", "with_peer", n.Address(), "from_round", last.Round+1)
+
+	for {
+		select {
+		case beaconPacket := <-beaconCh:
+			s.log.Debug("sync_manager", "new_beacon_fetched",
+				"with_peer", n.Address(),
+				"from_round", last.Round+1,
+				"got_round", beaconPacket.GetRound())
+			beacon := protoToBeacon(beaconPacket)
+
+			// verify the signature validity
+			if err := s.verifier.VerifyBeacon(*beacon, s.info.PublicKey); err != nil {
+				s.log.Debug("sync_manager", "invalid_beacon", "with_peer", n.Address(), "round", beacon.Round, "err", err, fmt.Sprintf("%+v", beacon))
+				return false
+			}
+
+			if err := s.store.Put(beacon); err != nil {
```

Comment:
We should probably have a common error to say if the store failed because of X reason (filesystem etc) OR because another round just got inserted before from elsewhere in the program (regular beacon loop) - in that case, it may not be necessary to continue the sync anymore.

Reply:
This is now tracked in #946 and left for a future PR.